File: 2024/day-12/main2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Point struct {
	x, y int
}

func readInputFile(filename string) [][]rune {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var grid [][]rune

	for scanner.Scan() {
		line := scanner.Text()
		grid = append(grid, []rune(line))
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	return grid
}

func isValid(x, y, rows, cols int) bool {
	return x >= 0 && x < rows && y >= 0 && y < cols
}

func bfs(grid [][]rune, visited [][]bool, startX, startY int) ([]Point, rune) {
	rows := len(grid)
	cols := len(grid[0])
	queue := []Point{{startX, startY}}
	regionCells := []Point{}
	plantType := grid[startX][startY]

	directions := []Point{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	visited[startX][startY] = true

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		regionCells = append(regionCells, current)

		for _, dir := range directions {
			nx, ny := current.x+dir.x, current.y+dir.y
			if isValid(nx, ny, rows, cols) && !visited[nx][ny] && grid[nx][ny] == plantType {
				queue = append(queue, Point{nx, ny})
				visited[nx][ny] = true
			}
		}
	}

	return regionCells, plantType
}

func calculateAreaAndSides(regionCells []Point, grid [][]rune) (int, int) {
	rows := len(grid)
	cols := len(grid[0])
	area := len(regionCells)
	sides := 0

	directions := []Point{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	for _, cell := range regionCells {
		for _, dir := range directions {
			nx, ny := cell.x+dir.x, cell.y+dir.y
			if !isValid(nx, ny, rows, cols) || grid[nx][ny] != grid[cell.x][cell.y] {
				sides++
			}
		}
	}

	return area, sides
}

func calculateFencingPrice(grid [][]rune) int {
	rows := len(grid)
	cols := len(grid[0])

	visited := make([][]bool, rows)
	for i := range visited {
		visited[i] = make([]bool, cols)
	}

	totalPrice := 0

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if !visited[i][j] {
				regionCells, _ := bfs(grid, visited, i, j)
				area, sides := calculateAreaAndSides(regionCells, grid)
				totalPrice += area * sides
			}
		}
	}

	return totalPrice
}

func main() {
	grid := readInputFile("input.txt")
	totalPrice := calculateFencingPrice(grid)
	fmt.Printf("Total price of fencing: %d\n", totalPrice)
}

File: 2024/day-12/main2.py
from collections import deque

def parse_map(grid):
    rows = len(grid)
    cols = len(grid[0])
    visited = [[False] * cols for _ in range(rows)]

    def in_bounds(x, y):
        return 0 <= x < rows and 0 <= y < cols

    def bfs(start_x, start_y, plant_type):
        queue = deque([(start_x, start_y)])
        visited[start_x][start_y] = True
        area = 0
        sides = 0

        while queue:
            x, y = queue.popleft()
            area += 1

            for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nx, ny = x + dx, y + dy
                if not in_bounds(nx, ny) or grid[nx][ny] != plant_type:
                    sides += 1
                elif not visited[nx][ny]:
                    visited[nx][ny] = True
                    queue.append((nx, ny))

        return area, sides

    regions = []
    for i in range(rows):
        for j in range(cols):
            if not visited[i][j]:
                region_type = grid[i][j]
                area, sides = bfs(i, j, region_type)
                regions.append((region_type, area, sides))

    return regions

def calculate_total_price(grid):
    regions = parse_map(grid)
    total_price = sum(area * sides for _, area, sides in regions)
    return total_price

# Read input from file
with open("input.txt", "r") as file:
    grid = [line.strip() for line in file]

total_price = calculate_total_price(grid)
print("Total Price:", total_price)

